Bring random-js back up to speed with newer webpack and node

rawr51919 opened the pull request:
All dependencies have been updated to their latest versions, linter errors have been fixed, build errors in webpack 5+ have been fixed, there are many things that have changed (even the author's credit got updated to her correct name, she is Summer now).
One unfortunate side effect of the updates is that ES3 support is dropped. @ckknight please let me know if there are any more things you want me to do before this gets merged, I'm sure I got everything covered.

Things left to do in this PR:

- [x] Add support for automatic dependency updates and automerging now and in the future using Mend Renovate (configuration is already in this PR, @ckknight will need to do some work on her end to enable this, namely adding the integration for Renovate to this repository and mayhaps her account too: https://github.com/apps/renovate) (once done, this will fix #66)
- [x] Rewire the build so it runs on GitHub Actions instead (fixing #67)
- [x] Add support for React Native (fixing #45 and #40)
- [x] ~~Include '.js' in imports (fixes #35) (May not be possible to fix this one due to rollup complaining about it)~~
- [x] Add `min` and `max` commands (fixing #42)
- [x] Add support for automatically choosing the best engine for platform being used (fixing #43) (As of this PR, this can be achieved through `const random = Random.auto();`)
- [x] Add documentation on using this library without Node and with Webpack 5+ (fixing #69 (nice), and #59)
- [x] Better documentation on making custom engines that interface with this library (fixing #68 and potentially paving the way for fixing #19)
- [x] (ties into the above, secondary objective) Build a xorgens-4096 PRNG for this engine (fixing #19)
- [x] Documentation on how to replicate `Math.random()` using this engine (fixing #38)
- [x] Better documentation on why `randomInt()` is biased (fixing #13 using info from https://github.com/ckknight/random-js/issues/13#issuecomment-143953826 and helping better explain why a library like this one is necessary)
- [x] Fix all warnings in `npm test` (secondary objective yet will be a lot cleaner when tests are run)
- [x] Add documentation about `random-js-no-node` and link to it somewhere in the README
- [x] Fix minor typos and documentation errors contained in various PRs (PRs #64 (courtesy of @aaby242abhay), #55 (courtesy of @MohammadXroid), #33 (courtesy of @pgerlach), and #22 (courtesy of @kennynaoh))

Fixes #69 
Fixes #68 
Fixes #67 
Fixes #64 
Closes #63
Closes #62 
Closes #61 
Closes #60 
Fixes #59 
Closes #58 
Closes #57 
Closes #56 
Fixes #55 
Closes #54 
Closes #53 
Closes #52 
Closes #50 
Closes #49 
Closes #48 
Closes #46 
Closes #45 
Fixes #43 
Fixes #42 
Fixes #40 
Closes #39 
Fixes #38 
Closes #35 
Fixes #33 
Fixes #22 
Fixes #13

rawr51919 committed:
Update dependencies and attempt a fix for newer versions of webpack

rawr51919 committed:
Update dependencies further and allow npm installs to finally work again

rawr51919 committed:
another npm install meme

rawr51919 committed:
npm install fixed for good

rawr51919 committed:
unnecessary ^ removals reverted

rawr51919 committed:
add lock file for bun (https://bun.sh)

rawr51919 committed:
remove unnecessary packages

rawr51919 committed:
fix the last of the critical problems

rawr51919 committed:
Create dependabot.yml

ckknight commented:
Oh wow, thanks for this!

> ES3 support is dropped

I'm really not sure if I care in 2023

> the author's credit got updated to her correct name

You're a gem 💎 

----
I'll take a further look this weekend and hopefully merge this in, as long as it doesn't introduce any unexpected breakages.

A large part of me wants to revamp the entire package, or split it into multiple pieces, or some other mechanism to make sure it's as easy as possible to tree-shake.

The best approach for that is probably to spruce it up while keeping it the same (read: this PR), then make decisions from there.

rawr51919 commented:
About the only thing left would be to implement #64, #55, #33, and #22, then all of random-js will be up to speed entirely

Will also be fixing #59, #40, #35, and #13 as a result of this PR, there is a bunch of warnings in 'npm test' I can't seem to figure out yet everything passes otherwise

rawr51919 commented:
> Oh wow, thanks for this!
> 
> > ES3 support is dropped
> 
> I'm really not sure if I care in 2023
> 
> > the author's credit got updated to her correct name
> 
> You're a gem 💎
> 
> I'll take a further look this weekend and hopefully merge this in, as long as it doesn't introduce any unexpected breakages.
> 
> A large part of me wants to revamp the entire package, or split it into multiple pieces, or some other mechanism to make sure it's as easy as possible to tree-shake.
> 
> The best approach for that is probably to spruce it up while keeping it the same (read: this PR), then make decisions from there.

This would likely become a sufficient base for version 3.0
Also, have Dependabot automatically merge dependency updates if the tests pass with the new dependency through either itself or GitHub Actions so the repo doesn't get clogged with dependency update PRs (issue created at #66)
Also the builds should get swapped to GitHub Actions as Travis CI is no longer a truly free build service (issue created at #67)

rawr51919 commented:
OK, I answered #38 and will try to close it with the merging of this PR once it's all finished

rawr51919 commented:
@ckknight I'm getting back onto this PR, was kinda hard to think about where to start without putting up a todo list of everything that needs done before it's ready for merge

rawr51919 committed:
typo fixed in README, lock files removed from package, and min/max commands added

rawr51919 committed:
Merge branch 'master' of https://github.com/rawr51919/random-js

rawr51919 commented:
Made sense to convert this to a draft, college work has gotten in the way, I'll be back to work on this hopefully soon

rawr51919 committed:
Add changes from various PRs to docs

rawr51919 committed:
enable automerge

thejustinwalsh commented:
This is still relevant in 2025, love this library, realistically what needs to be done to bring this up to date. How can I help?

rawr51919 commented:
> This is still relevant in 2025, love this library, realistically what needs to be done to bring this up to date. How can I help?

Basically look at which things I haven't checkmarked here yet and, if you want to, take a crack at them

rawr51919 committed:
Big update, lots has happened in here!

rawr51919 committed:
Fix tslint and fix markdownlint's config file

rawr51919 committed:
Condense build steps as prepublish is already ran after deps install

rawr51919 committed:
Moved the linting to ESLint and converted the Jest config to ESM

rawr51919 commented:
Big update, sorry for the 2-year-long wait! The Mend Renovate config is ready to use when @ckknight's got it set up here, the xorgens-4096 engine is implemented, builds have been moved to GitHub Actions (and are passing!), the TSLint build step has been migrated to ESLint + full ESM in the Jest config, automatic engine detection has been implemented (use `const random = Random.auto();` to use), added documentation on Webpack 5+ polyfills, React Native, and [random-js-no-node](https://www.npmjs.com/package/random-js-no-node), and added a new baseline document for adding new RNG engines to `random-js`.

Almost done, just a few more things left to do and this should be ready to merge.

@ckknight Can you please look through the new changes and tell me if I need to do anything?

rawr51919 committed:
Add information on why randomInt() is biased based on Summer's own comment

rawr51919 committed:
Baseline document for adding new RNG engines to `random-js` added

rawr51919 committed:
Added documentation on replicating `Math.random()` in this library

rawr51919 committed:
Add React Native documentation on Webpack 5+ polyfills (checks off the last checkmark!)

rawr51919 committed:
fix the last of the warnings

rawr51919 commented:
#35 is unfixable because it causes errors in Rollup when the final build is done after the tests.

Can someone please close that?

rawr51919 commented:
And with that, _ready to test, check, and merge!_

@Spacerat Your no-node branch might be serving a purpose now

rawr51919 committed:
readd yarn.lock

rawr51919 committed:
add bun.lock, update yarn.lock

rawr51919 committed:
Update the Build Status in the README to GitHub Actions too

rawr51919 committed:
Fix JS Markdown for `Math.random()` recreation

rawr51919 committed:
Fix typo in README

rawr51919 committed:
Add xorgens-4096 mention in readme

Spacerat commented:
Huh, I totally forgot I did that.

Am I understanding correctly that your docs are now going to point to that branch as the right way to use Random-JS on React Native?

I've not used or updated that branch for a long time - not since I created it. Do you think it’s OK of it remains as it is? I wonder if it makes sense for you to take ownership of it.

rawr51919 commented:
> 
> Huh, I totally forgot I did that.
> 
> Am I understanding correctly that your docs are now going to point to that branch as the right way to use Random-JS on React Native?
> 
> I've not used or updated that branch for a long time - not since I created it. Do you think it’s OK of it remains as it is? I wonder if it makes sense for you to take ownership of it.

The docs say either way, literally up to you to decide which package to use on React Native.

If you want me to take over the reins of that one I'd be happy to, I'd have to copypasta all the changes made in this PR and send them to that too although that's OK, will be waiting for Summer to merge it here first before the no-node version gets it merged too (or I could wait until this PR gets merged here and then sync the fork to match)

In the process I'd have to strip the rest of the nodeCrypto code completely out of the no-node version to remove unnecessary bloat in that package

Would it even be possible to take over the NPM package for no-node so new packages can be uploaded from my end?

rawr51919 committed:
fix typo in new 'adding RNG engines' document

rawr51919 committed:
add editorconfig

dependabot[bot] committed:
Bump js-yaml from 3.14.1 to 3.14.2

Bumps [js-yaml](https://github.com/nodeca/js-yaml) from 3.14.1 to 3.14.2.
- [Changelog](https://github.com/nodeca/js-yaml/blob/master/CHANGELOG.md)
- [Commits](https://github.com/nodeca/js-yaml/compare/3.14.1...3.14.2)

---
updated-dependencies:
- dependency-name: js-yaml
  dependency-version: 3.14.2
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump glob

Bumps  and [glob](https://github.com/isaacs/node-glob). These dependencies needed to be updated together.

Updates `glob` from 10.4.5 to 10.5.0
- [Changelog](https://github.com/isaacs/node-glob/blob/main/changelog.md)
- [Commits](https://github.com/isaacs/node-glob/compare/v10.4.5...v10.5.0)

Updates `glob` from 11.0.3 to 11.1.0
- [Changelog](https://github.com/isaacs/node-glob/blob/main/changelog.md)
- [Commits](https://github.com/isaacs/node-glob/compare/v10.4.5...v10.5.0)

---
updated-dependencies:
- dependency-name: glob
  dependency-version: 10.5.0
  dependency-type: indirect
- dependency-name: glob
  dependency-version: 11.1.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump lodash from 4.17.21 to 4.17.23

Bumps [lodash](https://github.com/lodash/lodash) from 4.17.21 to 4.17.23.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/compare/4.17.21...4.17.23)

---
updated-dependencies:
- dependency-name: lodash
  dependency-version: 4.17.23
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

codeCraft-Ritik reviewed (commented):
Great work! The implementation is clean and easy to understand

rawr51919 committed:
Fix typo in ADDING NEW RNG ENGINES.md

rawr51919 committed:
Merge pull request #3 from rawr51919/dependabot/npm_and_yarn/lodash-4.17.23

Bump lodash from 4.17.21 to 4.17.23

rawr51919 committed:
Merge pull request #2 from rawr51919/dependabot/npm_and_yarn/multi-12cb0b0c21

Bump glob

rawr51919 committed:
Merge pull request #1 from rawr51919/dependabot/npm_and_yarn/js-yaml-3.14.2

Bump js-yaml from 3.14.1 to 3.14.2
